Mathlib/Geometry/Manifold/ContMDiff/Product.lean

```diff
@@ -347,3 +347,156 @@ theorem contMDiff_pi_space :
+  apply Filter.mem_of_superset ?_ aux
+  rw [← I.image_eq (chartAt H x).target]
+  exact (chartAt H x).extend_image_target_mem_nhds (mem_chart_source _ x)
+
+-- TODO: move to ChartedSpace, sum section
+-- make some of these simp?
+lemma sum_chartAt_inl_apply {x : M} :
+    ((chartAt H (.inl x : M ⊕ M')) (Sum.inl x)) = (chartAt H x) x := by
+  rw [ChartedSpace.sum_chartAt_inl]
+  exact PartialHomeomorph.lift_openEmbedding_apply _ _
+
+lemma sum_chartAt_inr_apply {x : M'} :
+    ((chartAt H (.inr x : M ⊕ M')) (Sum.inr x)) = (chartAt H x) x := by
+  rw [ChartedSpace.sum_chartAt_inr]
+  exact PartialHomeomorph.lift_openEmbedding_apply _ _
+
+lemma extChartAt_inl_apply {x : M} :
+    ((extChartAt I (.inl x : M ⊕ M')) (Sum.inl x)) = (extChartAt I x) x := by
+  simp [sum_chartAt_inl_apply]
+
+lemma extChartAt_inr_apply {x : M'} :
+    ((extChartAt I (.inr x : M ⊕ M')) (Sum.inr x)) = (extChartAt I x) x := by
+  simp [sum_chartAt_inr_apply]
+
+lemma ContMDiff.sum_elim {f : M → N} {g : M' → N}
+    (hf : ContMDiff I J n f) (hg : ContMDiff I J n g) : ContMDiff I J n (Sum.elim f g) := by
+  intro p
+  rw [contMDiffAt_iff]
+  refine ⟨(Continuous.sum_elim hf.continuous hg.continuous).continuousAt, ?_⟩
+  cases p with
+  | inl x =>
+    -- In charts around x : M, the map f ⊔ g looks like f.
+    -- This is how they both look like in extended charts.
+    have : ContDiffWithinAt 𝕜 n ((extChartAt J (f x)) ∘ f ∘ (extChartAt I x).symm)
+        (range I) ((extChartAt I (.inl x : M ⊕ M')) (Sum.inl x)) := by
+      let hf' := hf x
+      rw [contMDiffAt_iff] at hf'
+      simpa only [extChartAt_inl_apply] using hf'.2
+    apply this.congr_of_eventuallyEq
+    · simp only [extChartAt, Sum.elim_inl, ChartedSpace.sum_chartAt_inl,
+        Sum.inl_injective.extend_apply]
+      filter_upwards with a
+      congr
+    · -- They agree at the image of x.
+      simp only [extChartAt, ChartedSpace.sum_chartAt_inl, Sum.elim_inl]
+      congr
+  | inr x =>
+    -- In charts around x : M, the map f ⊔ g looks like g.
+    -- This is how they both look like in extended charts.
+    have : ContDiffWithinAt 𝕜 n ((extChartAt J (g x)) ∘ g ∘ (extChartAt I x).symm)
+        (range I) ((extChartAt I (.inr x : M ⊕ M')) (Sum.inr x)) := by
+      let hg' := hg x
+      rw [contMDiffAt_iff] at hg'
+      simpa only [extChartAt_inr_apply] using hg'.2
+    apply this.congr_of_eventuallyEq
+    · simp only [extChartAt, Sum.elim_inl, ChartedSpace.sum_chartAt_inl,
+        Sum.inl_injective.extend_apply]
+      filter_upwards with a
+      congr
+    · -- They agree at the image of x.
+      simp only [extChartAt, ChartedSpace.sum_chartAt_inl, Sum.elim_inl]
+      congr
+
+variable [Nonempty H']
+
+lemma ContMDiff.sum_map {f : M → N} {g : M' → N'}
+    (hf : ContMDiff I J n f) (hg : ContMDiff I J n g) : ContMDiff I J n (Sum.map f g) :=
+  ContMDiff.sum_elim (ContMDiff.inl.comp hf) (ContMDiff.inr.comp hg)
+
+-- Better names welcome!
```

Comment:
The continuous analogue is called `continuous_sum_dom` and proves both the inl and inr statement. I'm not convinced that's better here.